src/scanpy/preprocessing/_pca/_dask_sparse.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Protocol, cast, overload
+
+import numpy as np
+import scipy.linalg
+from numpy.typing import NDArray
+
+from .._utils import _get_mean_var
+
+if TYPE_CHECKING:
+    from typing import Literal
+
+    from numpy.typing import DTypeLike
+    from scipy import sparse
+
+    from ..._compat import DaskArray
+
+    CSMatrix = sparse.csr_matrix | sparse.csc_matrix
+
+
+class PCASparseFit(Protocol):
+    n_components: int
+
+    n_components_: int
+    n_samples_: int
+    n_features_in_: int
+    dtype_: np.dtype
+    mean_: NDArray[np.floating]
+    components_: NDArray[np.floating]
+    explained_variance_: NDArray[np.floating]
+    explained_variance_ratio_: NDArray[np.floating]
+    noise_variance_: NDArray[np.floating]
+
+    def fit(self, x: DaskArray) -> PCASparseFit: ...
+    def transform(self, X: DaskArray) -> DaskArray: ...
+
+
+@dataclass
+class PCASparseDask:
+    n_components: int | None = None
+
+    def fit(self, x: DaskArray) -> PCASparseFit:
+        # this method makes `self` into the fitted version
+        self = cast(PCASparseFit, self)
+        assert isinstance(x.shape, tuple)
+        self.n_components_ = (
+            min(x.shape[:2]) if self.n_components is None else self.n_components
+        )
+        self.n_samples_ = x.shape[0]
+        self.n_features_in_ = x.shape[1] if x.ndim == 2 else 1
+        self.dtype_ = x.dtype
+        covariance, self.mean_ = _cov_sparse_dask(x)
+        self.explained_variance_, self.components_ = scipy.linalg.eigh(
+            covariance, lower=False
+        )
+
+        # Arrange eigenvectors and eigenvalues in descending order
+        self.explained_variance_ = self.explained_variance_[::-1]
+        self.components_ = np.flip(self.components_, axis=1)
+        self.components_ = self.components_.T[: self.n_components_, :]
+
+        self.explained_variance_ratio_ = self.explained_variance_ / np.sum(
+            self.explained_variance_
+        )
+        if self.n_components_ < min(self.n_samples_, self.n_features_in_):
+            self.noise_variance_ = self.explained_variance_[self.n_components_ :].mean()
+        else:
+            self.noise_variance_ = np.array([0.0])
+        self.explained_variance_ = self.explained_variance_[: self.n_components_]
+
+        self.explained_variance_ratio_ = self.explained_variance_ratio_[
+            : self.n_components_
+        ]
+        return self
+
+    def transform(self: PCASparseFit, x: DaskArray) -> DaskArray:
+        if TYPE_CHECKING:
```

Comment:
Why do we need this internally?

Reply:
because it makes my editor happy.

ideally someone would publish dask type hints or dask provides them instead of these hacks …

Reply:
Ok, I guess I was asking because we don't use type hints in this function from `da.`

Reply:
It doesn’t have any result at runtime, so are you OK with keeping it that way until the dask typing situation improves?

Reply:
Sure, no worries, I just wanted to understand.  If it's just "my computer wants it" and we don't know why, that's ok.  I don't even care to investigate.  Just wanted to learn :)

Reply:
Maybe best to comment in this case

Reply:
I’d say `if TYPE_CHECKING` is the comment here, it means “this is the version I want the type checker to see instead of the real code”, but OK, I’ll be more specific!